subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/IndexPageGenerator.java

```diff
@@ -16,96 +16,238 @@
+import java.io.UncheckedIOException;
 import java.io.Writer;
-import java.util.Date;
-import java.util.LinkedHashMap;
+import java.util.Comparator;
 import java.util.List;
-import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static java.util.Comparator.comparing;
+import static java.util.Comparator.comparingDouble;
 
 public class IndexPageGenerator extends HtmlPageGenerator<ResultsStore> {
-    private final List<NavigationItem> navigationItems;
+    public static final int DANGEROUS_REGRESSION_CONFIDENCE_THRESHOLD = 90;
+    private final Set<ScenarioBuildResultData> scenarios;
+    private final ResultsStore resultsStore;
+    private final String commitId = Git.current().getCommitId();
+
+    public IndexPageGenerator(ResultsStore resultsStore, File resultJson) {
+        this.resultsStore = resultsStore;
+        this.scenarios = readBuildResultData(resultJson);
+    }
+
+    private Set<ScenarioBuildResultData> readBuildResultData(File resultJson) {
+        try {
+            List<ScenarioBuildResultData> list = new ObjectMapper().readValue(resultJson, new TypeReference<List<ScenarioBuildResultData>>() {
+            });
+            return sortBuildResultData(list.stream().map(this::queryExecutionData));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @VisibleForTesting
+    Set<ScenarioBuildResultData> sortBuildResultData(Stream<ScenarioBuildResultData> data) {
+        Comparator<ScenarioBuildResultData> comparator = comparing(ScenarioBuildResultData::isBuildFailed).reversed()
+            .thenComparing(ScenarioBuildResultData::isSuccessful)
+            .thenComparing(comparing(ScenarioBuildResultData::isAboutToRegress).reversed())
+            .thenComparing(comparingDouble(ScenarioBuildResultData::getConfidencePercentage).reversed())
+            .thenComparing(comparingDouble(ScenarioBuildResultData::getRegressionPercentage).reversed())
+            .thenComparing(ScenarioBuildResultData::getScenarioName);
+        return data.collect(() -> new TreeSet<>(comparator), TreeSet::add, TreeSet::addAll);
+    }
+
+    private ScenarioBuildResultData queryExecutionData(ScenarioBuildResultData scenario) {
+        PerformanceTestHistory history = resultsStore.getTestResults(scenario.getScenarioName(), 3, 2, ResultsStoreHelper.determineChannel());
+        List<? extends PerformanceTestExecution> recentExecutions = history.getExecutions();
+        List<? extends PerformanceTestExecution> currentCommitExecutions = recentExecutions.stream().filter(execution -> execution.getVcsCommits().contains(commitId)).collect(Collectors.toList());
+        if (currentCommitExecutions.isEmpty()) {
+            scenario.setRecentExecutions(recentExecutions.stream().map(this::extractExecutionData).collect(Collectors.toList()));
```

Comment:
Can we filter here by the last commit, so we only have changes for one commit in the list?

Reply:
Yes, we can but I think current is better. If `currentCommitExecutions` is empty, we present the recent 3 executions to users to let them decide which one is the right commit (since there're exeuction time) in case we miss something (the last commit is not the right one which the user expects).